File: apps/vsc-plugin/src/services/pairing.service.ts
import * as vscode from 'vscode';
import * as os from 'os';
import * as http from 'http';
import * as https from 'https';
import * as crypto from 'crypto';
import { PROTOCOL_VERSION } from '@codeam/shared';
import { SettingsService, type RecentSession } from './settings.service';
import { CommandRelayService } from './command-relay.service';
import { checkApiReachable } from './connectivity';
import { OutputChannel } from 'vscode';

export type PairingCodeResult = { code: string; expiresAt: number } | { blocked: true } | null;

export interface PairedUserInfo {
  name: string;
  email: string;
  plan: string;
  currentPeriodEnd?: string;
}

export interface PairingListener {
  onPaired(sessionId: string): void;
}

export class PairingService {
  private static instance: PairingService;
  private pollInterval: NodeJS.Timeout | null = null;
  private pollTimeout: NodeJS.Timeout | null = null;
  private listeners: PairingListener[] = [];
  private log: OutputChannel;

  pairedUser: PairedUserInfo | null = null;
  currentSessionId: string | null = null;

  private constructor(log: OutputChannel) {
    this.log = log;
  }

  static initialize(log: OutputChannel): PairingService {
    PairingService.instance = new PairingService(log);
    return PairingService.instance;
  }

  static getInstance(): PairingService {
    if (!PairingService.instance) {
      throw new Error('PairingService not initialized');
    }
    return PairingService.instance;
  }

  addListener(listener: PairingListener): void {
    this.listeners.push(listener);
  }

  async requestPairingCode(): Promise<PairingCodeResult> {
    const settings = SettingsService.getInstance();
    const pluginId = settings.ensurePluginId();
    const relay = CommandRelayService.getInstance();

    // Preflight: if the API host is unreachable (VPN/firewall/allowlist),
    // surface the cloud fallback instead of a cryptic pairing failure.
    if ((await checkApiReachable(settings.apiBaseUrl)) === 'blocked') {
      this.log.appendLine('[pairing] API unreachable (preflight) — offering cloud fallback');
      return { blocked: true };
    }

    let branch: string | null = null;
    try {
      const { ProjectOpsService } = await import('./project-ops.service');
      branch = await ProjectOpsService.detectCurrentBranch();
    } catch { /* branch unknown */ }

    try {
      const result = await relay.postJson(`${settings.apiBaseUrl}/api/pairing/code`, {
        pluginId,
        ideName: 'VS Code',
        ideVersion: vscode.version,
        hostname: os.hostname(),
        branch,
        // SEC crit1 (#813): enroll the PoP hash so /status + /reconnect
        // require this window's secret. Older backends ignore it.
        pluginSecretHash: crypto
          .createHash('sha256')
          .update(settings.ensurePollSecret())
          .digest('hex'),
      });

      if (result?.data) {
        const data = result.data as Record<string, unknown>;
        const code = data.code;
        const expiresAt = data.expiresAt;
        if (typeof code === 'string' && typeof expiresAt === 'number') {
          this.startPollingForPairing();
          return { code, expiresAt };
        }
      }
      return null;
    } catch (e) {
      // postJson rejects ONLY on transport error/timeout (it resolves on any
      // HTTP response), so a throw here means no HTTP response = blocked.
      this.log.appendLine(`[pairing] request failed (transport) — offering cloud fallback: ${e}`);
      return { blocked: true };
    }
  }

  stopPolling(): void {
    if (this.pollInterval) {
      clearInterval(this.pollInterval);
      this.pollInterval = null;
    }
    if (this.pollTimeout) {
      clearTimeout(this.pollTimeout);
      this.pollTimeout = null;
    }
  }

  private startPollingForPairing(): void {
    this.stopPolling();
    this.pollInterval = setInterval(() => this.checkPairingStatus(), 3000);
    this.pollTimeout = setTimeout(() => this.stopPolling(), 300_000);
  }

  private async checkPairingStatus(): Promise<void> {
    const settings = SettingsService.getInstance();
    const pluginId = settings.ensurePluginId();

    try {
      // SEC crit1 (#813): replay this window's PoP secret so the gated
      // /status returns the token + PII (the VSC plugin reads the token
      // from here). Legacy backends ignore the header.
      const result = await this.getJson(
        `${settings.apiBaseUrl}/api/pairing/status?pluginId=${pluginId}`,
        { 'X-Plugin-Poll-Secret': settings.ensurePollSecret() },
      );

      if (result?.data) {
        const data = result.data as Record<string, unknown>;
        const paired = data.paired as boolean;

        if (paired) {
          const sessionId = data.sessionId as string;
          const userObj = data.user as Record<string, unknown> | undefined;

          if (userObj) {
            this.pairedUser = {
              name: (userObj.name as string) || '',
              email: (userObj.email as string) || '',
              plan: (userObj.plan as string) || 'FREE',
              currentPeriodEnd: userObj.currentPeriodEnd as string | undefined,
            };
          }

          // Persist the per-pairing token. Replayed as
          // `X-Plugin-Auth-Token` on every authed call so we still
          // pass auth after the legacy fallback expires (2026-05-25).
          const rawToken = data.pluginAuthToken;
          if (typeof rawToken === 'string' && rawToken.length > 0) {
            SettingsService.getInstance().setPluginAuthToken(rawToken);
            this.log.appendLine(
              `[pairing] Stored pluginAuthToken (${rawToken.length} chars)`,
            );
            // The new token clears any prior 401 — re-arm the
            // "session expired" toast for the next failure.
            CommandRelayService.getInstance().resetAuthFailureGate();
          } else {
            // Diagnostic — earlier reports of "[fileWatcher] no
            // pluginAuthToken" trace back to here. Surface the actual
            // shape we got so we can tell whether the backend stopped
            // returning the field, returned it under a different
            // name, or sent something non-string.
            this.log.appendLine(
              `[pairing] WARN no pluginAuthToken on status response — rawToken=${typeof rawToken} keys=${Object.keys(data).join(',')}`,
            );
          }

          this.currentSessionId = sessionId;
          this.log.appendLine(`Pairing detected! Session: ${sessionId}, user: ${this.pairedUser?.email}`);
          this.stopPolling();
          this.saveCurrentSession();
          this.listeners.forEach((l) => l.onPaired(sessionId));
        }
      }
    } catch (e) {
      // Silent polling error
    }
  }

  private saveCurrentSession(): void {
    if (!this.currentSessionId || !this.pairedUser) { return; }
    const settings = SettingsService.getInstance();
    settings.addRecentSession({
      sessionId: this.currentSessionId,
      userName: this.pairedUser.name,
      userEmail: this.pairedUser.email,
      userPlan: this.pairedUser.plan,
      connectedAt: Date.now(),
    });
  }

  clearCurrentSession(): void {
    this.currentSessionId = null;
    this.pairedUser = null;
    SettingsService.getInstance().setPluginAuthToken(null);
  }

  /**
   * Re-attach a saved paired session without requiring a fresh QR code.
   *
   * This deliberately avoids CommandRelayService.postJson(): that helper adds
   * the cached X-Plugin-Auth-Token to every request, and a stale token can make
   * /api/pairing/reconnect return 401 before it has a chance to mint the fresh
   * token we need. The reconnect endpoint trusts the sessionId+pluginId tuple
   * plus the poll secret, matching the CLI boot path.
   */
  async reconnectSession(
    sessionId: string,
    cached?: Partial<RecentSession>,
  ): Promise<boolean> {
    const settings = SettingsService.getInstance();
    const pluginId = settings.ensurePluginId();

    try {
      const result = await this.postJson(
        `${settings.apiBaseUrl}/api/pairing/reconnect`,
        { pluginId, sessionId },
        { 'X-Plugin-Poll-Secret': settings.ensurePollSecret() },
      );

      const success = (result as Record<string, unknown> | null)?.success === true;
      if (!success) return false;

      const data = (result as Record<string, unknown>).data as Record<string, unknown> | undefined;
      const rawToken = data?.pluginAuthToken;
      if (typeof rawToken === 'string' && rawToken.length > 0) {
        settings.setPluginAuthToken(rawToken);
        this.log.appendLine(`[reconnect] Stored pluginAuthToken (${rawToken.length} chars)`);
        CommandRelayService.getInstance().resetAuthFailureGate();
      } else {
        this.log.appendLine(
          `[reconnect] WARN no pluginAuthToken on reconnect response — rawToken=${typeof rawToken}`,
        );
      }

      const userObj = data?.user as Record<string, unknown> | undefined;
      this.onReconnected(sessionId, {
        name: (userObj?.name as string) || cached?.userName || '',
        email: (userObj?.email as string) || cached?.userEmail || '',
        plan: (userObj?.plan as string) || cached?.userPlan || 'FREE',
        currentPeriodEnd: userObj?.currentPeriodEnd as string | undefined,
      });
      return true;
    } catch (e) {
      this.log.appendLine(`Reconnect error: ${e}`);
      return false;
    }
  }

  onReconnected(sessionId: string, user: PairedUserInfo): void {
    this.currentSessionId = sessionId;
    this.pairedUser = user;
    this.saveCurrentSession();
    this.listeners.forEach((l) => l.onPaired(sessionId));
  }

  private async getJson(
    url: string,
    extraHeaders?: Record<string, string>,
  ): Promise<Record<string, unknown> | null> {
    return new Promise((resolve, reject) => {
      const urlObj = new URL(url);
      const transport = urlObj.protocol === 'https:' ? https : http;
      const req = transport.request(
        {
          hostname: urlObj.hostname,
          port: urlObj.port,
          path: urlObj.pathname + urlObj.search,
          method: 'GET',
          ...(extraHeaders ? { headers: extraHeaders } : {}),
          timeout: 10000,
        },
        (res: http.IncomingMessage) => {
          let body = '';
          res.on('data', (chunk: Buffer) => { body += chunk.toString(); });
          res.on('end', () => {
            try { resolve(JSON.parse(body)); } catch { resolve(null); }
          });
        },
      );
      req.on('error', reject);
      req.on('timeout', () => { req.destroy(); reject(new Error('timeout')); });
      req.end();
    });
  }

  private async postJson(
    url: string,
    body: Record<string, unknown>,
    extraHeaders?: Record<string, string>,
  ): Promise<Record<string, unknown> | null> {
    return new Promise((resolve, reject) => {
      const data = JSON.stringify(body);
      const urlObj = new URL(url);
      const transport = urlObj.protocol === 'https:' ? https : http;
      const req = transport.request(
        {
          hostname: urlObj.hostname,
          port: urlObj.port,
          path: urlObj.pathname + urlObj.search,
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'Content-Length': Buffer.byteLength(data),
            'X-Codeam-Protocol-Version': PROTOCOL_VERSION,
            ...(extraHeaders ?? {}),
          },
          timeout: 10000,
        },
        (res: http.IncomingMessage) => {
          let responseBody = '';
          res.on('data', (chunk: Buffer) => { responseBody += chunk.toString(); });
          res.on('end', () => {
            try { resolve(JSON.parse(responseBody)); } catch { resolve(null); }
          });
        },
      );
      req.on('error', reject);
      req.on('timeout', () => { req.destroy(); reject(new Error('timeout')); });
      req.write(data);
      req.end();
    });
  }
}

File: apps/jetbrains-plugin/src/main/kotlin/com/windsurf/controller/services/PairingService.kt
package com.windsurf.controller.services

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.Service
import com.intellij.openapi.diagnostic.Logger
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.net.InetAddress
import java.util.Timer
import java.util.TimerTask
import java.util.concurrent.TimeUnit

@Service(Service.Level.APP)
class PairingService {

    private val logger = Logger.getInstance(PairingService::class.java)
    private val gson = Gson()
    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build()
    private var pollingTimer: Timer? = null

    /** Back-to-back throttled/failed polls; drives the exponential backoff. */
    private var consecutiveThrottles = 0

    /** Epoch ms at which this pairing attempt gives up (replaces a leaked Timer). */
    private var pollingDeadlineAt = 0L
    data class PairedUserInfo(
        val name: String,
        val email: String,
        val plan: String,
        val currentPeriodEnd: String? = null
    )

    interface PairingListener {
        fun onPaired(sessionId: String)
    }

    var pairedUser: PairedUserInfo? = null
        private set

    var currentSessionId: String? = null
        private set

    private val listeners = mutableListOf<PairingListener>()

    fun addListener(listener: PairingListener) {
        listeners.add(listener)
    }

    /**
     * Sealed result for [requestPairingCode]:
     * - [Code]    — successful code returned by the backend.
     * - [Blocked] — API host unreachable (preflight or transport error / no HTTP response).
     *               Callers should offer the cloud-fallback panel.
     * - [None]    — backend returned a non-2xx HTTP status (real server error; treat as
     *               the existing null path — do NOT show the cloud-fallback panel).
     */
    sealed interface PairingCodeResult {
        data class Code(val code: String, val expiresAt: Long) : PairingCodeResult
        data object Blocked : PairingCodeResult
        data object None : PairingCodeResult
    }

    /**
     * Request a pairing code from the backend.
     *
     * Classification (mirrors VS Code pairing.service.ts):
     *  • Preflight BLOCKED  → [PairingCodeResult.Blocked]  (no HTTP attempt)
     *  • Transport error    → [PairingCodeResult.Blocked]  (IOException/timeout — no HTTP response)
     *  • Non-2xx HTTP status → [PairingCodeResult.None]    (real server error — unchanged path)
     *  • 2xx success        → [PairingCodeResult.Code]
     *
     * Must be called from a background thread (OkHttp blocks; do NOT call on the EDT).
     */
    fun requestPairingCode(): PairingCodeResult {
        val settings = SettingsService.getInstance()

        // Preflight: if the API host is unreachable (VPN/firewall/allowlist),
        // surface the cloud fallback instead of a cryptic pairing failure.
        val reachability = ConnectivityChecker.checkApiReachable(settings.state.apiBaseUrl)
        if (reachability == Reachability.BLOCKED) {
            logger.info("[pairing] API unreachable (preflight) — offering cloud fallback")
            return PairingCodeResult.Blocked
        }

        val pluginId = settings.ensurePluginId()

        val body = JsonObject().apply {
            addProperty("pluginId", pluginId)
            addProperty("ideName", "WebStorm")
            addProperty("ideVersion", com.intellij.openapi.application.ApplicationInfo.getInstance().fullVersion)
            addProperty("hostname", getHostname())
            // SEC crit1 (#813): enroll the PoP hash so /status + /reconnect
            // require this install's secret. Older backends ignore it.
            addProperty("pluginSecretHash", settings.pollSecretHash())
            // ⚠️ Repo y rama, como los envian el CLI y el plugin de VS Code.
            //
            // El backend acepta `branch` desde hace tiempo y las apps titulan la
            // sesion con el repo — pero este plugin no enviaba ninguno de los
            // dos, PESE A CALCULARLOS: `detectRepoSlug()` y `gitStatus()` ya
            // existian y solo se usaban para redactar el texto de un dialogo de
            // fallback. Resultado: en la lista de sesiones, toda sesion de
            // JetBrains salia sin proyecto y sin rama, y habia que titularla por
            // la herramienta.
            //
            // Best-effort: si el proyecto no es un repo git, o no tiene remoto
            // de GitHub, se omiten y el emparejamiento sigue igual que antes.
            runCatching {
                val ops = ProjectOpsService.getInstance()
                ops.detectRepoSlug()?.let {
                    addProperty("repoIdentifier", "${'$'}{it.owner}/${'$'}{it.repo}")
                }
                ops.gitStatus()
                    .get("branch")
                    ?.takeIf { !it.isJsonNull }
                    ?.asString
                    ?.takeIf { it.isNotBlank() && it != "(detached)" }
                    ?.let { addProperty("branch", it) }
            }.onFailure {
                logger.info("[pairing] no git context for this project: ${'$'}{it.message}")
            }
        }

        val request = Request.Builder()
            .url("${settings.state.apiBaseUrl}/api/pairing/code")
            .post(gson.toJson(body).toRequestBody("application/json".toMediaType()))
            .build()

        return try {
            val response = httpClient.newCall(request).execute()
            val responseBody = response.body?.string() ?: return PairingCodeResult.None

            if (response.isSuccessful) {
                val json = gson.fromJson(responseBody, JsonObject::class.java)
                val data = json.getAsJsonObject("data")
                val result = PairingCodeResult.Code(
                    code = data.get("code").asString,
                    expiresAt = data.get("expiresAt").asLong
                )
                startPollingForPairing()
                result
            } else {
                // Non-2xx HTTP response — real server error (NOT a connectivity block).
                logger.warn("Failed to get pairing code: $responseBody")
                PairingCodeResult.None
            }
        } catch (e: IOException) {
            // Transport error: no HTTP response received (connection refused, timeout, DNS
            // failure, etc.) — classify as Blocked so the caller shows the cloud-fallback
            // panel. This mirrors VS Code's postJson reject path.
            logger.info("[pairing] request failed (transport) — offering cloud fallback: ${e.message}")
            PairingCodeResult.Blocked
        } catch (e: Exception) {
            // Any other unexpected error — treat as a real error, not a connectivity block.
            logger.error("Error requesting pairing code", e)
            PairingCodeResult.None
        }
    }

    fun stopPolling() {
        pollingTimer?.cancel()
        pollingTimer = null
    }

    private fun startPollingForPairing() {
        stopPolling()
        consecutiveThrottles = 0
        pollingDeadlineAt = System.currentTimeMillis() + POLL_WINDOW_MS
        pollingTimer = Timer("pairing-poll", true)
        schedulePoll(2000)
    }

    /**
     * Schedule ONE poll, `delayMs` from now.
     *
     * Deliberately a single-shot `schedule` that re-arms itself rather than
     * `scheduleAtFixedRate`: the HTTP call below is synchronous inside the
     * TimerTask, and fixed-RATE scheduling fires catch-up executions
     * back-to-back whenever a response is slower than the period — so a slow or
     * rate-limited backend turned this loop into a burst generator. Fixed
     * DELAY also lets each poll pick its own wait, which is what honouring
     * `Retry-After` requires.
     *
     * The old code additionally leaked one `Timer` per pairing attempt for the
     * 5-minute cutoff; the deadline is now just a timestamp checked here.
     */
    private fun schedulePoll(delayMs: Long) {
        val timer = pollingTimer ?: return
        if (System.currentTimeMillis() >= pollingDeadlineAt) {
            stopPolling()
            return
        }
        try {
            timer.schedule(object : TimerTask() {
                override fun run() {
                    val outcome = checkPairingStatus()
                    if (outcome == null) return // paired → polling already stopped
                    consecutiveThrottles =
                        if (outcome.throttled) consecutiveThrottles + 1 else 0
                    schedulePoll(
                        nextPollDelayMs(
                            status = outcome.status,
                            retryAfterSeconds = outcome.retryAfterSeconds,
                            consecutiveThrottles = consecutiveThrottles,
                        ),
                    )
                }
            }, delayMs)
        } catch (e: IllegalStateException) {
            // Timer was cancelled between the null-check and the schedule.
            logger.debug("[pairing] poll timer already cancelled: ${e.message}")
        }
    }

    /** One poll's outcome, or null once pairing completed (loop stopped). */
    private data class PollOutcome(
        val status: Int,
        val retryAfterSeconds: Int?,
        val throttled: Boolean,
    )

    private fun checkPairingStatus(): PollOutcome? {
        val settings = SettingsService.getInstance()
        val pluginId = settings.ensurePluginId()

        val request = Request.Builder()
            .url("${settings.state.apiBaseUrl}/api/pairing/status?pluginId=$pluginId")
            // SEC crit1 (#813): replay the PoP secret so the gated /status
            // returns the token + PII (the plugin reads the token here).
            .header("X-Plugin-Poll-Secret", settings.ensurePollSecret())
            .get()
            .build()

        try {
            val response = httpClient.newCall(request).execute()
            val status = response.code
            // `Retry-After` is what the backend's throttler sends with a 429.
            // Ignoring it is exactly how this loop turned into a request flood.
            val retryAfterSeconds = response.header("Retry-After")?.trim()?.toIntOrNull()
            val throttled = status == 429 || status >= 500
            if (throttled) {
                response.close()
                logger.debug("[pairing] status poll throttled ($status) — backing off")
                return PollOutcome(status, retryAfterSeconds, throttled = true)
            }
            val body = response.body?.string()
                ?: return PollOutcome(status, retryAfterSeconds, throttled = false)

            if (response.isSuccessful) {
                val json = gson.fromJson(body, JsonObject::class.java)
                val data = json.getAsJsonObject("data")
                val paired = data.get("paired").asBoolean
                if (paired) {
                    val sessionId = data.get("sessionId").asString
                    val userObj = data.getAsJsonObject("user")
                    if (userObj != null) {
                        pairedUser = PairedUserInfo(
                            name = userObj.get("name")?.asString ?: "",
                            email = userObj.get("email")?.asString ?: "",
                            plan = userObj.get("plan")?.asString ?: "FREE",
                            currentPeriodEnd = userObj.get("currentPeriodEnd")?.takeIf { !it.isJsonNull }?.asString
                        )
                    }
                    // Persist the per-pairing token. Replayed as
                    // `X-Plugin-Auth-Token` on every authed call so
                    // we still pass auth after the legacy fallback
                    // expires (2026-05-25).
                    val rawToken = data.get("pluginAuthToken")?.takeIf { !it.isJsonNull }?.asString
                    if (!rawToken.isNullOrEmpty()) {
                        SettingsService.getInstance().setPluginAuthToken(rawToken)
                        // The new token clears any prior 401 — re-arm the
                        // "session expired" notification gate for the next failure,
                        // and un-latch the output publisher (a fresh pairing
                        // supersedes the dead one it latched on).
                        CommandRelayService.getInstance().resetAuthFailureGate()
                        AgentOutputPublisher.resetPairingLatch()
                    }
                    currentSessionId = sessionId
                    logger.info("Pairing detected! Session: $sessionId, user: ${pairedUser?.email}")
                    stopPolling()
                    saveCurrentSession()
                    listeners.forEach { it.onPaired(sessionId) }
                    return null
                }
            }
            return PollOutcome(status, retryAfterSeconds, throttled = false)
        } catch (e: Exception) {
            // Transport failure (no HTTP response): status 0 → treated as
            // throttle-worthy so a dead network doesn't spin at full speed.
            logger.debug("Polling error: ${e.message}")
            return PollOutcome(status = 0, retryAfterSeconds = null, throttled = true)
        }
    }

    private fun saveCurrentSession() {
        val sid = currentSessionId ?: return
        val user = pairedUser ?: return
        val settings = SettingsService.getInstance()
        settings.addRecentSession(SettingsService.RecentSession(
            sessionId = sid,
            userName = user.name,
            userEmail = user.email,
            userPlan = user.plan,
            connectedAt = System.currentTimeMillis()
        ))
    }

    fun clearCurrentSession() {
        currentSessionId = null
        pairedUser = null
        SettingsService.getInstance().setPluginAuthToken(null)
        // Tear down the Path B file watcher - it would otherwise keep
        // POSTing to a session that no longer exists. The service is
        // idempotent on stop(), safe to call when never started.
        try {
            FileWatcherService.getInstance().stop()
        } catch (e: Exception) {
            logger.debug("Failed to stop FileWatcherService on unpair: ${e.message}")
        }
    }

    fun onReconnected(sessionId: String, user: PairedUserInfo) {
        currentSessionId = sessionId
        pairedUser = user
        saveCurrentSession()
        listeners.forEach { it.onPaired(sessionId) }
    }

    private fun getHostname(): String {
        return try {
            InetAddress.getLocalHost().hostName
        } catch (e: Exception) {
            "unknown"
        }
    }

    companion object {
        /** How long one pairing attempt keeps polling before giving up. */
        private const val POLL_WINDOW_MS = 300_000L

        fun getInstance(): PairingService =
            ApplicationManager.getApplication().getService(PairingService::class.java)
    }
}
